File: set7/5/main.cpp
#include<iostream>
#include<queue>
#include<tuple>
#include<vector>
#include<algorithm>

using namespace std;

vector<tuple<long long, long long, long long, int>> circle;
vector<vector<int>> tree;

pair<int, int> BFS(int S) {
    pair<int, int> ret;
    queue<pair<int, int>> Q;
    vector<int> used(tree.size());
    Q.emplace(S, 0);
    used[S] = 1;
    while(!Q.empty()) {
        auto [cur, d] = Q.front(); Q.pop();
        if(ret.first < d) {
            ret.first = d;
            ret.second = cur;
        }
        for(int nxt : tree[cur]) {
            if(used[nxt]) continue;
            used[nxt] = true;
            Q.emplace(nxt, d + 1);
        }
    }
    return ret;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N; cin >> N;
    tree.resize(N + 1);
    for(int i = 0; i < N; i++) {
        int x, y, r; cin >> x >> y >> r;
        circle.emplace_back(r, x, y, i + 1);
    }
    sort(circle.begin(), circle.end());
    circle.emplace_back(1000000000, 0, 0, 0);
    for(int i = 0; i <= N; i++) {
        for(int j = i + 1; j <= N; j++) {
            auto [ar, ax, ay, aidx] = circle[i];
            auto [br, bx, by, bidx] = circle[j];
            long long mr = ar - br;
            long long d  = (ax - bx) * (ax - bx) + (ay - by) * (ay - by);
            if((d == 0 && mr != 0) || d < mr * mr) {
                tree[aidx].push_back(bidx);
                tree[bidx].push_back(aidx);
                break;
            }
        }
    }

    auto cur = BFS(0);
    auto cur2 = BFS(cur.second);
    cout << cur2.first;

    return 0;
}

File: set7/5/main.py
import sys
from collections import deque, defaultdict

def input():
    return sys.stdin.readline().rstrip()

def BFS(S):
    ret = [0, 0]
    Q = deque([(S, 0)])
    used = [0] * (N + 1)
    used[S] = 1
    while Q:
        cur, d = Q[0]
        Q.popleft()
        if ret[0] < d:
            ret = [d, cur]
        for nxt in tree[cur]:
            if used[nxt]: continue
            used[nxt] = True
            Q.append((nxt, d + 1))
    return ret

N = int(input())
tree = defaultdict(list)

circle = []
for i in range(N):
    x, y, r = map(int, input().split())
    circle.append((r, x, y, i + 1))
circle = sorted(circle)
circle.append((1000000000, 0, 0, 0))
for i in range(N + 1):
    for j in range(i + 1, N + 1):
        ar, ax, ay, aidx = circle[i]
        br, bx, by, bidx = circle[j]
        mr, d = ar - br, (ax - bx) * (ax - bx) + (ay - by) * (ay - by)
        if (d == 0 and mr != 0) or d < mr * mr:
            tree[aidx].append(bidx)
            tree[bidx].append(aidx)
            break

print(BFS(BFS(0)[1])[0])